test/e2e_federation/util.go

```diff
@@ -150,28 +139,82 @@ func createService(clientset *fedclientset.Clientset, namespace, name string) (*
 	return clientset.Services(namespace).Create(service)
 }
 
+func createService(clientset *fedclientset.Clientset, namespace, name string) (*v1.Service, error) {
+	if clientset == nil || len(namespace) == 0 {
+		return nil, fmt.Errorf("Internal error: invalid parameters passed to createService: clientset: %v, namespace: %v", clientset, namespace)
+	}
+	By(fmt.Sprintf("Creating federated service %q in namespace %q", name, namespace))
+
+	// Tests can be run in parallel, so we need a different nodePort for
+	// each test.
+	// we add in a array all the "available" ports
+	availablePorts := make([]int32, FederatedSvcNodePortLast-FederatedSvcNodePortFirst)
+	for i := range availablePorts {
+		availablePorts[i] = int32(FederatedSvcNodePortFirst + i)
+	}
+
+	var err error
+	// until the availablePort list is not empty, lets try to create the service
+	for len(availablePorts) > 0 {
+		// select the Id of an available port
+		i := rand.Intn(len(availablePorts))
+
+		service, err := createServiceWithNodePort(clientset, namespace, name, availablePorts[i])
+		if err == nil {
+			return service, nil
+		}
+
+		// check if service have been created in some clusters.
+		// if it's the case, delete them.
+		if found, _ := hasService(clientset, namespace, name); found {
```

Comment:
Here clientset is for federation api server. I think what you wanted was to check if service is created in all federated clusters with the given NodePort and if the service creation in any one of federated clusters has failed, then probably delete that service in federation and ensure that the service is deleted in all federated cluster and create a service with another NodePort. And probably keep doing this until the service is created in all federated clusters with the chosen NodePort.